File: minDepth.cpp
#include <queue>
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    int minDepth(TreeNode *root)
    {
        if (root == nullptr)
        {
            return 0;
        }

        std::queue<TreeNode *> q;                 
        q.push(root);
        TreeNode *b = root;
        TreeNode *p = root;
        int n = 1;
        while (!q.empty())
        {
            p = q.front();
            q.pop();
            if (p->left == nullptr && p->right == nullptr)
            {
                break;
            }
            if (p->left != nullptr)
            {
                q.push(p->left);
            }
            if (p->right != nullptr)
            {
                q.push(p->right);
            }

            if (p == b)
            {
                n += 1;
                b = q.back();
            }
        }
        return n;
    }

    int getDepth(TreeNode *root, int n)
    {

        if (root->left == nullptr && root->right == nullptr)
        {
            return n;
        }
        int a = 100000, b = 100000;
        if (root->left != nullptr)
        {
            a = getDepth(root->left, n + 1);
        }

        if (root->right != nullptr)
        {
            b = getDepth(root->right, n + 1);
        }

        return std::min(a, b);
    }

    int minDepth2(TreeNode *root)
    {
        if (root == nullptr)
        {
            return 0;
        }

        return getDepth(root, 1);
    }
};

File: JavaCodeSpace/src/cn/medemede/leecode/MinDepth.java
package cn.medemede.leecode;

import cn.medemede.leecode.modules.TreeNode;

import java.util.LinkedList;

public class MinDepth {

    /**
     * 二叉树最小深度
     * <p>广度优先遍历BFS，遇到叶子返回结果</p>
     *
     * @param root
     * @return
     */
    public int minDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int res = 0;
        LinkedList<TreeNode> queue = new LinkedList<>();
        queue.addLast(root);
        TreeNode last = root;
        while (!queue.isEmpty()) {
            TreeNode tmp = queue.pollFirst();
            if (tmp.left == null && tmp.right == null) {
                res++;
                return res;
            }
            if (tmp.left != null) {
                queue.addLast(tmp.left);
            }
            if (tmp.right != null) {
                queue.addLast(tmp.right);
            }
            if (tmp == last) {
                res++;
                last = queue.peekLast();
            }
        }
        return res;
    }


}

